optuna/study/_optimize.py

```diff
@@ -220,6 +220,8 @@ def _run_trial(
         frozen_trial = study._storage.get_trial(trial._trial_id)
         raise
     finally:
+        # We must invalidate all trials cache here as it is only valid within a trial.
+        study._thread_local.cached_all_trials = None
```

Comment:
This is not the right place to invalidate the cache. The stale cache will be used forever if optuna is used from ask and tell interface.